### ISSUE TITLE
Error Source for HTTP Headers

### ISSUE THREAD

**Issue by fny** [json-api/json-api#1514]
Error sources as currently defined do not provide a mechanism to specify errors resulting from invalid HTTP headers. Adding a `header` field would satisfy this use case:

> * `source`: an object containing references to the source of the error,
>   optionally including any of the following members:
>   * `pointer`: a JSON Pointer [[RFC6901](https://tools.ietf.org/html/rfc6901)]
>     to the value in the request document that caused the error [e.g. `"/data"`
>     for a primary data object, or `"/data/attributes/title"` for a specific
>     attribute]. This **MUST** point to a value in the request document that
>     exists; if it doesn't, the client **SHOULD** simply ignore the pointer.
>   * `parameter`: a string indicating which URI query parameter caused
>     the error.
>   * `header`: a string indicating which HTTP header caused
>     the error.

**Comment by gabesullice** [comment]
Hi! Thanks for the suggestion. Can you provide an example/use case?

My gut reaction is that this makes sense. However, let me play devil's advocate.

My first, less important concern: Aren't error objects already too complicated to process? By adding this, we add yet another object member which may or may not be there and we therefore increase the complexity of client implementations. 

The more concrete concern: The spec only uses three headers: `accept`, `content-type` and `location`, but clients only send the `accept` and `content-type` headers. There are already dedicated HTTP status codes for errors in those headers: `406 Not Acceptable` and `415 Unsupported Media Type`. Why do we need an affordance for headers not already covered by these status codes?

**Comment by fny** [comment]
Thanks for the push back: it's always interesting to consider alternative views.

> The spec only uses three headers

Query parameters outside of the spec, for example, can be referenced properly, and there's no restriction on referencing errors for out-of-spec query parameters. Say I require a "lang" param. Easy to encode as en error. Now say I require "Accept-Language". I send a 422 and point to what exactly?

It's just a little strange that the only part of the entire HTTP/1.1 request that can't be referenced are the headers.

I do think there's another issue lurking here that your concern surfaces.  Should errors on `accept`, `content-type` or `location` have sources? This seems cumbersome.

Here is another concrete use case. Say I require an authentication token  (Authentication Bearer) and XSRF token for each request.

Now say I want to tell the client that the XSRF token is wrong, expired, or requires a refresh. I want to send them an error with details, but based on the current spec, I can't point that header as the issue. Is this a 401 because of the authentication token or the XSRF token? Currently, I use an error code specific to that error as the distinction. Say there are other errors in the request too. Why should all the other errors have a source and not this error?

I know that use case is rather niche. The most common outside use case I can think of would be similar issues which cause a single 401 or 403 error. Do we just assume the code itself is suffices as a reference? I'll admit my case is somewhat pathological, however, it is real.

Work arounds do exist, though: there is meta after all.

Anyway, I'll ponder this more. Would be great to hear other thoughts too. In my heart, I just want to be able to reference all the things.

**Comment by gabesullice** [comment]
I hear your point about being able to point to query parameters but not headers. It's a fair point.

> In my heart, I just want to be able to reference all the things.

Makes sense!

> I want to send them an error with details, but based on the current spec, I can't point that header as the issue. Is this a 401 because of the authentication token or the XSRF token?

This is beside the point, but I personally wouldn't send a `401` for an invalid XSRF token because a) that token isn't really a credential and b) `401` responses comes with [other, unrelated responsibilities](https://tools.ietf.org/html/rfc7235#section-3.1).

XSRF tokens are more of a request "validation" and in that sense, the [`400` language](https://httpstatuses.com/400) seems more appropriate.

However, `400` is an even _more_ ambiguous status and so I think that it actually bolsters the case for a `header` member.

<hr>

I think I'm now more in support of this addition than against it. I can see how status codes are too ambiguous. I'm curious what @dgeb thinks.

**Comment by freddrake** [comment]
My take, FWIW:

When a request is received and the endpoint is identified using the URL from the request, content negotiation is used to determine the appropriate response format.  When an error occurs, that changes what is being returned to the error information, and content negotiation should then be applied to determine the appropriate format.  At this point, JSON:API (or RFC 7807, or ...) may be selected, even though the request is not a JSON:API request.

At this point, it's entirely reasonable to support indicating that the error content comes from a header.  Some applications may require or support non-standard headers as well (especially for legacy pre-JSON:API endpoints), making it reasonable to report bad headers.

**Comment by richard-rance** [comment]
[OpenAPI defines](http://spec.openapis.org/oas/v3.0.3#parameter-locations) `query`, `header`, `path`, and `cookie` all as types of parameters. Instead of adding a new member for each would we want to add a `location` member? It could also have `body` as a valid value.

**Comment by freddrake** [comment]
Consolidating in `location` would then require a way to provide both the location name and location type.  The current structure allows more than one to be involved; it might make more sense to think about whether *that* makes sense for a single error.  Aside from that, which way to arrange it feels like a six-of-one v. half-dozen-of-the-other comparison; which is more convenient depends more on your implementation language & frameworks.

**Comment by richard-rance** [comment]
To clarify I was envisioning both the `location` and either `parameter` or `pointer` members to be used at the same time. Something like this:

> - `source`: an object containing references to the source of the error, optionally including any of the following members:
>   - `pointer`: a JSON Pointer [RFC6901] to the value in the request document that caused the error [e.g. "/data" for a primary data object, or "/data/attributes/title" for a specific attribute]. This MUST point to a value in the request document that exists; if it doesn’t, the client SHOULD simply ignore the pointer.
>   - `parameter`: a string indicating which request parameter caused the error.
>   - `location`: A string indicating which part of the request contained the error. The recommended values are `query`, `header`, `path`, `cookie`, and `body`. When not provided a location of `query` can be assumed if the `parameter` member is populated or `body` if the `pointer` member is populated.

The current format discourages an error from having multiple sources. If there are two query params that are not compatible the service would need to respond with an error for each param. Similarly, if the validation of a member in the body is affected by the value of another member you can only point to one of the body members per error. 

An oddity of the current format is it allows you to specify both a point in the body and a query param as the source of the error. I'm not sure of a justification for that which would not also justify multiple query params. 

We could extend @fny's proposal with one member per location with the addition that only a single member of the source should be set. Which option is more convenient would indeed be a matter of opinion.

**Comment by freddrake** [comment]
@richard-rance: It sounds like we're largely in agreement.  Whether the location is provided as a separate property or implied by the name of the property holding the reference isn't particularly important to me.

Clarifying that the `source` object can contain at most one member (or two, if `location` is added), would be an improvement.  Requiring it to contain exactly one (or two, with `location`) would be even better.  There could be backward compatibility issues if `location` were required, though.

**Comment by gabesullice** [comment]
Thanks for the great discussion here! :)

I agree with @freddrake about clarifying the contents of the `source` member, but I'm not sure if the rule should be `>0`, `=1`, or `<2`. Let's reason through it...

An empty `source` member would be pointless. In that case, the `source` member should be omitted altogether. I think that eliminates `<2`.

The current text doesn't prevent one from using custom members, nor from having multiple members. That means it could be a BC break to specify `=1` for some implementations. Therefore, we're left with `>0`.

So, I think we should clarify that the `source` member, if present, **MUST** contain at least one member.

There's one last thing I think we could also clarify in a relatively BC way... I think we could make the optional members mutually exclusive. In other words, say that you can't have `pointer` && `parameter` at the same time.

---

@richard-rance, I like your thinking :)

I think you're bringing up two important questions:

#### 1. Do we want to consolidate to something like OpenAPI's `in` member so that we don't have to have a unique member for each location?
I think the advantage of the `in`/`location` idea is that it is a little more future-friendly because you can add new allowed values for the `location` member. It feels like a smaller change to add new allowed values to an existing member than it does to add new members entirely. In practice, client implementations will be hardcoded to expect whatever is allowed, in either case though, so it's really a matter of taste and if we didn't already have `pointer` and `parameter` I would probably have preferred `location` too. 

Since we didn't start with something like `location`, let's just add a `header` member as a sibling of `parameter` instead of a new member that duplicates some of what we already have.

Still, what about OpenAPI's `query`, `header`, `path`, and `cookie`? Should we have equivalents for every one of them? I don't think so.

* `query`: this is the same as `parameter` IMO
* `header`: yes, let's add this one.
* `path`: I think a `404` and a `detail` are enough here. Also, I'd rather not encourage APIs to use complex path parameters.
* `cookie`: This is a little redundant, isn't it? Implementations can specify `"header": "cookie"`
* `body`: this is the same as `pointer`. Unlike OpenAPI, we know the format of the request body will always be JSON and so we can use the more capable JSON pointer spec.

#### 2. What to do about errors caused by the interaction of different fields/parameters?
I think our errors objects were probably designed for form submissions (I'm not sure, wasn't there), maybe not explicitly, but it was probably on everyone's mind. If that was the case, I think `pointer` was really about helping the client put the error message in the "right" place in the form. If the error was specific to a particular form field, the pointer told the client to render a red box over a particular form field.

If two fields interact, I think you wouldn't want to show the same error twice. So, you'd either put the error at the top of the form (IOW, omit `source` entirely) or use the `pointer` to point to the most appropriate field for the error to be displayed.

That's an Ember/React/Angular specific reasoning, but I think it applies generally... You can think about it like this: if the first query parameter restricts the use of a second query parameter, the error is that the second parameter was added despite the fact it was no longer permitted. Thus, the source of the error is the second, restricted parameter instead of the source of the error being _both_ parameters.

**Comment by gabesullice** [comment]
Whoops, probably should add a _tl;dr_...

**tl;dr:** Let's add a `header` member. Also, let's specify that `source` **MUST** contain at least one member and that it **MUST NOT** contain more than one of the following members: `pointer`, `parameter`, `header`

**Comment by freddrake** [comment]
> So, I think we should clarify that the `source` member, if present, **MUST** contain at least one member.

I'd generally prefer this for any sub-objects that have optional members.

Unfortunately, this is backward incompatible, since a JSON:API 1.0 response could legally return ``"source": {}``, even though it would be useless.  For 1.1, we strengthened the status member to **SHOULD**; perhaps we could do the same here, stating that an empty source **SHOULD** be omitted.

> There's one last thing I think we could also clarify in a relatively BC way... I think we could make the optional members mutually exclusive. In other words, say that you can't have `pointer` && `parameter` at the same time.

Again, I like this, but... the backward compatibility issue is a question.  It's hard to imagine anyone using these together in practice, but it would have been legal in 1.0.  The compatibility promise in the first paragraph of 1.0 is quite strong.

> Still, what about OpenAPI's `query`, `header`, `path`, and `cookie`? Should we have equivalents for every one of them? I don't think so.
> 
> * `query`: this is the same as `parameter` IMO
> * `header`: yes, let's add this one.
> * `path`: I think a `404` and a `detail` are enough here. Also, I'd rather not encourage APIs to use complex path parameters.

JSON:API also explicitly stays away from the URL structure.  If URLs are always handed to the client, the client should not be generating URLs anyway.  So this doesn't seem useful w/r/t JSON:API.

> * `cookie`: This is a little redundant, isn't it? Implementations can specify `"header": "cookie"`

Maybe, but ``{"cookie": "some-cookie-name"}`` is much more specific than ``{"header": "cookie"}``.  It's unlikely to make a difference for display to a user, but much better when debugging.

> * `body`: this is the same as `pointer`. Unlike OpenAPI, we know the format of the request body will always be JSON and so we can use the more capable JSON pointer spec.

@gabesullice: Did you see my comments on content negotiation in https://github.com/json-api/json-api/issues/1514#issuecomment-700128091?  Are you expecting JSON:API errors to only be generated from JSON:API requests?

**Comment by gabesullice** [comment]
Fair points wrt to backwards compatibility. I'm not opposed to **SHOULD**  and **SHOULD NOT**.

Re. `cookie`: good catch. I was not thinking about specific cookie names... I'm skeptical about its usefulness. Isn't a developer going to read the `detail` in the end? If we added `cookie-name` why would we not add `parameter-index` for parameters that accept multiple values (e.g. `include=foo,bar,unknown` => `parameter: include; parameter-index: 2`)?

I kind of wish we took an approach more like RFC 7807. I like that it is very generic and allows implementors to add their own members for their own purposes.

> Did you see my comments on content negotiation in #1514 (comment)? Are you expecting JSON:API errors to only be generated from JSON:API requests?

I did, but I didn't make this connection.  I don't think we can prevent someone from using JSON:API for responses to non-JSON:API requests, but our introduction does say:

> JSON:API is a specification for how a client should request that resources be fetched or modified, and how a server should respond to those requests.

Which makes me believe our focus should be on the conversation between JSON:API clients and servers, not on providing a universal JSON media type.

**Comment by freddrake** [comment]
> Re. `cookie`: good catch. I was not thinking about specific cookie names... I'm skeptical about its usefulness. Isn't a developer going to read the `detail` in the end? If we added `cookie-name` why would we not add `parameter-index` for parameters that accept multiple values (e.g. `include=foo,bar,unknown` => `parameter: include; parameter-index: 2`)?

I used to think people would read... but I've dealt with more of them now.

While I don't want to hold a developer's hands when reporting an error to them, making an error report easy to consume is a good way to help them find their error.  If it requires deciphering text in a natural language they may not be proficient at, it will eventually become a support request.  Possibly in a natural language I'm not proficient at.

> I kind of wish we took an approach more like RFC 7807. I like that it is very generic and allows implementors to add their own members for their own purposes.

There are definitely some good things there.  What I do like about the JSON:API error object is that there's a defined place for implementation-specific fields in meta.

> I think our focus should be on the conversation between JSON:API clients and servers, not on providing a universal JSON media type.

Yes, but.  In a world where content negotiation is a part of every interaction, I'd still rather get error messages in something better suited to debugging than anything else.

**Comment by richard-rance** [comment]
> If we added cookie-name why would we not add parameter-index for parameters that accept multiple values (e.g. include=foo,bar,unknown => parameter: include; parameter-index: 2)?

A cookie name is generally displayed by browsers and most frameworks as a separate key/value pair from the headers. The index in a list of values is interesting. It is more like a pointer within the parameter. Having just an index would not work for query params that are more complex such as an SQL-like filter. Possibly a `value` member might be a way to state what is wrong but even that feels implementation-specific. 

For myself building an error for your example the `details` would be `The value "unknown" is not a valid relationship to be included.` I would then put any of the dynamic parts of the string into the meta object so they can be worked on programmatically and are not affected by translations. 

> JSON:API also explicitly stays away from the URL structure. If URLs are always handed to the client, the client should not be generating URLs anyway. So this doesn't seem useful w/r/t JSON:API.

The server supporting HATEOAS and the client actually using it are two different things. Often clients want to jump directly to fetching their record without the discovery step. Telling them that they got a 404 because the 2nd path parameter is not valid could be useful.

**Comment by freddrake** [comment]
The notion of "cookie" as an error location, regardless of how it's encoded, suggests that both ``"cookie": "<name>"`` and ``"header": "cookie"`` could reasonably co-exist.  This isn't a concern as long as ``"cookie"`` is not a defined property for ``"source"``.

@gabesullice: Do you plan on proposing a specific change to the specification language, or would you prefer I create a PR?

**Comment by freddrake** [comment]
@gabesullice: PR #1533 makes the specific changes you outlined above.

**Comment by freddrake** [comment]
Should be closed now that PR #1533 has landed.
